### PR TITLE
[WIP] Module component

### PR DESCRIPTION
**Pull request by paulstelian97**
There is at least one use case where the code required to process audio data cannot be made open source (due to IP concerns), so a loadable module interface will be useful.

There is prior related work, namely in branch topic/module. The implementation from there only does a module_init/module_exit, and in the module_init the module can do anything. In my own work I instead create a new component type, and this component will have its functionality vary at runtime depending on the actual need.

The advantage to the new approach is that there is no longer an immediate need to add functionality into the ASoC or ALSA subsystems in the Linux kernel to support reloading topologies at runtime in this use case. Instead, the driver itself will load a SOF module (if not already loaded), register it with this component and then configure the component to use the functionality in the loaded module. The topology may thus still be defined statically while allowing this partial flexibility.

[WIP] because I haven't yet figured out how to make the interface of this functionality usable (how or where to change the ABI) and the actual way the module itself must be loaded.

Requires ABI change (minor ABI update).

Signed-off-by: Paul Olaru <paul.olaru@nxp.com>

### CODE REVIEW

**Comment by dbaluta** [comment]
Thanks @paulstelian97 for sending this RFC. This is useful to open an early discussion about "module" component type implementation.

Few general comments:

1) please use tags in commit message subject. I mean something like:
 -  audio: Implement loadable component functionality (draft) 

Hope that @plbossart and @lgirdwood can join the discussion from now on:

2) Lets try to better describe the solution. We need to have a pipeline like this:

-> pcm0 -> buffer -> module -> buffer -> dai

OR

<-pcm  <-buffer <- module <- buffer <- dai

OR (this is also very important for us for in memory encoding/decoding).

-> pcm0 -> buffer -> module
                                            /
 <- pcm1                   <- buffer

In the case where module is an decoder algorithm for example
it means that Linux host sends compressed data (e.g mp3) and gets
back decompressed pcm data.

After the pipeline is created we need to find a way to dynamically load the
binary of the algorithm (e.g mp3) and maybe later to be able to load another
algorithm (e.g aac).

I think that after the pipeline is created the type of the module to be loaded
and maybe the path to the binary can be specified inside set_params function,
and the actual loading of the binary to happen perhaps in with the .prepare
command of a trigger.

File: src/include/sof/audio/module.h [hunk 1]
```diff
@@ -29,11 +29,17 @@
 #define trace_module_error(__e, ...) \
 	trace_error(TRACE_CLASS_MODULE, __e, ##__VA_ARGS__)
 #define tracev_module(__e, ...) \
-	tracev_event(TRACE_CLASS_MUX, __e, ##__VA_ARGS__)
+	tracev_event(TRACE_CLASS_MODULE, __e, ##__VA_ARGS__)
```
Comment (dbaluta) [inline src/include/sof/audio/module.h]: Please do not modify this here. Please squash it in the first patch.

File: src/audio/comp_module/module.c [hunk 2]
```diff
@@ -0,0 +1,191 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2019 NXP Semiconductors. All rights reserved.
+//
```
Comment (dbaluta) [inline src/audio/comp_module/module.c]: This should be: Copyright 2019 NXP

**Comment by juimonen** [comment]
@paulstelian97 @dbaluta sorry can you still clarify the difference when using "module" compared to "process" component from linux driver side? We already have use cases of using "process" as a wrapper for closed proprietary code...

**Comment by paulstelian97** [comment]
From my understanding they are similar. However, unlike the Linux side, this module component supports changing its functionality without reloading the topology. Given that currently we don't support dynamically generating a topology every time we play something new, and that I don't think having tens of pipelines (one per format) would be practical (especially when only some are popular), I think this may still be useful. As such having a small amount of topologies (where the pipeline layout itself differs) and then having module components may allow more flexibility than currently available.

**Comment by juimonen** [comment]
@paulstelian97 hmm ok. So now when reading @dbaluta explanation more carefully I kind of see what you mean. I'm just not totally sure you get any benefit over reloading new topology: If we consider for example the case of changing decoder from mp3 to aac:
- I guess you are thinking that it is too much memory consumption to include all decoders into same component? 
- So you anyway need to have the logic somewhere to parse the correct format then send the actual runnable code binary to dsp?
- Or is the idea that the alternatives are preserved in kernel side perhaps?

I could see a case where you have multiple pipelines running and then you want to change the decoder (like without distracting simultaneous recording for example), so this is where module could be useful. Otherwise we could just reload new topology. 

I'm kind of worried about sending actual runnable code from userspace to dsp during runtime, OTH we are sending the whole binary during boot. 

Sorry for the braindump/fart, just trying to understand what you are after here...

**Comment by paulstelian97** [comment]
@juimonen Well I'm not sure you can even _have_ all the code from different decoders/encoders in the same component, due to licensing issues. The components are after all proprietary and made by different companies, so I don't think they would collaborate to make a single component having all formats. And even though we may have some open source ones also available, it would be rather weird to have the open source ones integrated in the firmware while the closed source ones remain separate.

From the point of view of actual memory occupied, it may be a concern on some systems (having ten decoders may inflate the occupied instruction memory significantly, enough to not have any room left). I will add some interface to unload such unused modules for this purpose.

We can send only trusted code, from a system path such as /lib/firmware like the firmware itself. The Linux driver would check the request, see what codec is requested and then request loading the module from said trusted path. So there isn't an actual problem from the security point of view. Whoever has write access to those paths must actually have root on the Linux system, and in that case whatever security we would have by avoiding sending runnable code to the DSP would be nullified anyway. Said alternatives may also be preserved in the kernel itself if that's more reliable.

In short, the kernel side will parse the module (located somewhere in /lib/firmware), then upload the sections to the DSP and trigger some sort of init code (which should just register this module). Then we will also configure the component itself to use the module (for the corresponding format) during the prepare() trigger, and also configure the module itself. Whenever the pipeline is idle/at the next prepare() trigger another module may be loaded and the current one unloaded (changing decoders, for example).

(I'm also open to discussion whether we should call this "module" or something else)

**Comment by dbaluta** [comment]
@juimonen thanks a lot for jumping in! Actually in the beginning my initial thought was to create a topology for each decoder type and just reload the topology whenever we need to. Anyhow, I think @paulstelian97 solution is more cleaner and elegant. And it is also more flexible:
- we just create one pipeline with a generic "module" component and load each instance of module at runtime most likely with a set_params type of call
- we need to support at least 5 running instances of encoding/decoding in parallel and we also have at least 15 codec types as of now. Loading everything from the start will waste a lot of memory.
- with @paulstelian97  approach we will only need to create one tplg file instead of 15

There are a lot of things left unclear as of now, but we are exploring every possible solution with advantages and drawbacks.

**Comment by juimonen** [comment]
@dbaluta @paulstelian97 yes thanks for the explanations. So I see this is targeted for 2 issues:
- possibly not enough memory in dsp to hold all decoder/processing variants
- possibility to dynamically change the pipeline code during runtime
 
Personally I see your point, however we need @lgirdwood and @tlauda to chime in. If you get some real demo working with dynamically loading and exchanging module code, would be really nice and even better basis for discussion. Anyhow, sounds interesting!

**Comment by keyonjie** [comment]
Hi @dbaluta and @paulstelian97 This topic is interesting. Some points here:

1. So here what the 'Module' component do is actually decoding? If so, rename it to decoder may be good for understanding.

2. We can already support dynamic configure blob, please refer to audio/detect_test.c and sof/tools/ctl/ctl.c.

3. If the intention is to load dynamic FW binaries, it might need add framework support firstly, which including both the SOF and the Linux driver side.

**Comment by paulstelian97** [comment]
> Hi @dbaluta and @paulstelian97 This topic is interesting. Some points here:
> 
> 1. So here what the 'Module' component do is actually decoding? If so, rename it to decoder may be good for understanding.

It can do decoding (on playback pipelines) or encoding (on record pipelines).

> 2. We can already support dynamic configure blob, please refer to audio/detect_test.c and sof/tools/ctl/ctl.c.

I'll look into them soon enough.

> 3. If the intention is to load dynamic FW binaries, it might need add framework support firstly, which including both the SOF and the Linux driver side.

Agreed, however I wanted to add a bit of code to decide on an interface before adding this support. I suspect some work from the branch `topic/module` that I mentioned in my first comment to this PR may be useful, but I'm still looking into it.

File: src/audio/comp_module/module.c [hunk 3]
```diff
@@ -0,0 +1,329 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2019 NXP. All rights reserved.
+//
+// Author: Paul Olaru <paul.olaru@nxp.com>
+
+#include <config.h>
+
+#if CONFIG_COMP_MODULE
```
Comment (jajanusz) [inline src/audio/comp_module/module.c]: Better just don't compile file at all if you don't need this feature.
f.e. something like that in cmake file->
```
if(CONFIG_COMP_MODULE)
	add_local_sources(sof module.c)
endif()
```

File: src/audio/comp_module/module.c [hunk 4]
```diff
@@ -0,0 +1,329 @@
...
+
+#include <stdint.h>
+#include <stddef.h> // what is it used for?
```
Comment (jajanusz) [inline src/audio/comp_module/module.c]: There may be a lot files included, because they were needed in the past and no1 removed them. If you have no need for the header just remove it. However remember that the fact that it compiles without something doesn't mean that it shouldn't be included explicitly. F.e. if you need stdint types you should include it instead of depending of something else including it.

File: src/audio/comp_module/module.c [hunk 5]
```diff
@@ -0,0 +1,329 @@
...
+#include <sof/stream.h> // What is this used for?
+#include <sof/audio/component.h>
+#include <sof/ipc.h> // What is this used for?
+#include <sof/audio/module.h>
+
+static struct comp_dev *module_new(struct sof_ipc_comp *comp)
+{
+	struct sof_ipc_comp_process *ipc_process =
+		(struct sof_ipc_comp_process *)comp;
+	//size_t bs = ipc_process->size;
```
Comment (jajanusz) [inline src/audio/comp_module/module.c]: Please don't leave commented code. Just remove it if it is not needed.

File: src/audio/modules/passthrough.c [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+	(void)data;
+	(void)max_data_size;
+	return 0;
+}
+
+static void passthrough_copy_16(struct comp_dev *dev,
+				struct comp_buffer *source,
+				struct comp_buffer *sink,
+				uint32_t frames
+			       )
```
Comment (jajanusz) [inline src/audio/modules/passthrough.c]: Checkpatch didn't complain about this lone parenthesis?

File: src/include/user/trace.h [hunk 7]
```diff
@@ -58,6 +58,8 @@ struct system_time {
 #define TRACE_CLASS_SELECTOR	(29 << 24)
 #define TRACE_CLASS_SCHEDULE	(30 << 24)
 #define TRACE_CLASS_SCHEDULE_LL	(31 << 24)
+/* 32 and 33 missing, trace classes should match between user and sof */
```
Comment (jajanusz) [inline src/include/user/trace.h]: It's mistake that it's in 2 places, I'll fix it.

File: src/audio/Kconfig [hunk 8]
```diff
@@ -44,6 +44,12 @@ config COMP_MIXER
 	help
 	  Select for Mixer component
 
+config COMP_MODULE
+	bool "Loadable module component"
+	default y
+	help
+	  Select for Module component
```
Comment (jajanusz) [inline src/audio/Kconfig]: Please give more verbose `help`. I'm not sure what I'm enabling here - capability to load modules? One or more? Static/dynamic?

File: src/audio/comp_module/module.c (hunk 4 above)
Comment (paulstelian97) [reply src/audio/comp_module/module.c]: Addressed, coming up in next push.

File: src/audio/comp_module/module.c (hunk 3 above)
Comment (paulstelian97) [reply src/audio/comp_module/module.c]: Thanks, I actually had a form of double-guarding (CMakeFiles AND in-source). Addressed.

File: src/audio/comp_module/module.c (hunk 5 above)
Comment (paulstelian97) [reply src/audio/comp_module/module.c]: Partially addressed. For module_cmd I still need to figure out how to interface with the userspace to decide upon which module to use at some point and on its configuration data.

File: src/audio/modules/passthrough.c (hunk 6 above)
Comment (paulstelian97) [reply src/audio/modules/passthrough.c]: I may not have some Git hooks configured. Addressed and will look into that.

File: src/audio/Kconfig (hunk 8 above)
Comment (paulstelian97) [reply src/audio/Kconfig]: I'll leave it as-is for now. This "module" component supports replacing functionality at runtime. While it was intended to be dynamic at the beginning it actually supports both static and dynamic modules. Any suggestions? (every module component can bind to one module at a time)

Comment (paulstelian97) [reply src/audio/Kconfig]: Added now.

**Comment by paulstelian97** [comment]
@keyonjie Thank you for the suggestion with the dynamic configure blob, I have ended up using that interface to decide at runtime which module in particular implements the functionality of this component (in the last push as of this writing)

File: src/audio/comp_module/module.c (hunk 5 above)
Comment (paulstelian97) [reply src/audio/comp_module/module.c]: Addressed with last push.

File: src/audio/comp_module/module.c [hunk 9]
```diff
@@ -0,0 +1,342 @@
...
+		mod = find_module_by_type(req_type);
+		ret = 0;
+		if (!mod) {
+			ret = -ENOENT;
+			break;
+		}
+		ret = set_module_ops(dev, mod);
+		if (ret < 0)
+			break;
+		// Create the instance of the selected module by means
```
Comment (dbaluta) [inline src/audio/comp_module/module.c]: Please use C style comments:
`` /* */ instead of //

**Comment by paulstelian97** [comment]
> At a high level we are missing locking around list manipulation and reference counting.

Will work on that once everything is cleared up. I wonder if we could make CI actually test for SMP correctness (the platform I'm working on has a single Xtensa core and any such bugs would not manifest).

> I'm also not following the use case here where the we can replace with component in a pipeline with another without reloading a new topology (since the new component may well have different kcontrols or scheduling needs that could impact the pipeline and driver) ?

For the use case @dbaluta mentioned that we may have multiple pipelines in the topology and doing a topology for every configuration would either be inflexible (all pipelines having the same kind of component), or would blow up exponentially (5 pipelines/instances with 14 different choices would mean over half a million topologies). Or it could mean that we have to generate the new topology on-device (not exactly a pleasant idea I suppose) AND disrupt existing active pipelines when we reload the topology.

Also, the component types would be limited. This module component is not intended to be an endpoint or to bring its own scheduling (hardware DAIs would do that or, in the loopback case, the recording endpoint). For components with kcontrols, they should have a compatible interface to support such replacement.

> We should be able to tear down a pipeline today and replace it with a new pipeline using existing code, but today we cannot modify a completed pipeline (without a teardown and rebuild due to userspace and driver pipeline coherency).

Would we be able to do this with a pipeline without removing others? Say that we have two PCM interfaces and two DAIs, and two independent pipelines connecting each PCM with its own DAI. Let's say that PCM0 is playing and needs to be replaced (of course we will stop and perhaps reset it first), and PCM1 is also playing and needs to remain like that. Is there a way to replace the pipeline from PCM0 without disrupting the other one with current code, with the pipeline to be replaced being stopped?

> An alternative would be for the topology to define all pipeline combinations needed by the usecases and then only load the correct pipeline for each use case on demand. i.e. userspace tells driver which pipelines to realize or destroy. This way we would have coherency between userspace, driver and FW (and we can change more than just a single component).

Addressed above. The question is about having two simultaneous pipelines, possibly with the same components (of course with different IDs) and a maximum of, say, 5 active pipelines.

**Comment by dbaluta** [comment]
> Apologies for the delay, just back to work. At a high level we are missing locking around list manipulation and reference counting.
> I'm also not following the use case here where the we can replace with component in a pipeline with another without reloading a new topology (since the new component may well have different kcontrols or scheduling needs that could impact the pipeline and driver) ?

The use case we are targeting is playing compressed audio. We need to support around 14 codec types (mp3, wma, aac, etc).

The compress API allows you to set the codec type before actually playing the samples via the compressed ALSA device. So, our thought was that when the codec type is set we just reload the module pointing the the correct decoding algorithm.

**Comment by lgirdwood** [comment]
> Would we be able to do this with a pipeline without removing others? Say that we have two PCM interfaces and two DAIs, and two independent pipelines connecting each PCM with its own DAI. Let's say that PCM0 is playing and needs to be replaced (of course we will stop and perhaps reset it first), and PCM1 is also playing and needs to remain like that. Is there a way to replace the pipeline from PCM0 without disrupting the other one with current code, with the pipeline to be replaced being stopped?

Thanks for the use case clarification, we could certainly add a new IPC to allow modification of a pipeline. i.e.
1) Pipeline constructed at init and used for MP3 decode, but use case now changed to AAC.
2) Userspace initiates change (new ioctl ?) and driver sends PIPELINE_MODIFY IPC.
3) Userspace / driver then delete MP3 component and insert AAC component using existing IPCs.
4) Driver then sends PIPELINE_COMPLETE IPC.
5) AAC decode can now start.

This flow would keep coherency between all FW/SW layers alongside the creation/destruction of any new kcontrols. However, this needs userspace code, maybe simpler (given the usecase to)

1) Same as 1 above.
2) Userspace uses kcontrol enum to change from MP3 to AAC (stream is in stopped state already).
3) "module" component then sends new IPC to host asking for new component module code (send me the AAC code).
4) Driver loads AAC module from host OS and sends to DSP.
5) AAC is ready.

Does the latter proposal make sense ? It would be simpler faster but will need a new IPC and a method to load the new code.

**Comment by paulstelian97** [comment]
The second approach does sound interesting and might be the way to go.

**Comment by lgirdwood** [comment]
@paulstelian97 any update ? I've marked for v1.6, but we can always push back to later.

**Comment by paulstelian97** [comment]
> @paulstelian97 any update ? I've marked for v1.6, but we can always push back to later.

I will let you know when I will focus on it, for now I prefer this is not marked as anything. Right now I'm on a design phase and the approach outlined in this PR may actually be changed significantly. I expect sometime March to actually have some activity on this (could be a little sooner)

**Comment by muddogxp** [comment]
> > Would we be able to do this with a pipeline without removing others? Say that we have two PCM interfaces and two DAIs, and two independent pipelines connecting each PCM with its own DAI. Let's say that PCM0 is playing and needs to be replaced (of course we will stop and perhaps reset it first), and PCM1 is also playing and needs to remain like that. Is there a way to replace the pipeline from PCM0 without disrupting the other one with current code, with the pipeline to be replaced being stopped?
> 
> Thanks for the use case clarification, we could certainly add a new IPC to allow modification of a pipeline. i.e.
> 
> 1. Pipeline constructed at init and used for MP3 decode, but use case now changed to AAC.
> 2. Userspace initiates change (new ioctl ?) and driver sends PIPELINE_MODIFY IPC.
> 3. Userspace / driver then delete MP3 component and insert AAC component using existing IPCs.
> 4. Driver then sends PIPELINE_COMPLETE IPC.
> 5. AAC decode can now start.
> 
> This flow would keep coherency between all FW/SW layers alongside the creation/destruction of any new kcontrols. However, this needs userspace code, maybe simpler (given the usecase to)
> 
> 1. Same as 1 above.
> 2. Userspace uses kcontrol enum to change from MP3 to AAC (stream is in stopped state already).
> 3. "module" component then sends new IPC to host asking for new component module code (send me the AAC code).
> 4. Driver loads AAC module from host OS and sends to DSP.
> 5. AAC is ready.
> 
> Does the latter proposal make sense ? It would be simpler faster but will need a new IPC and a method to load the new code.

@lgirdwood do we have any example topology or module code support compressed -> decompressed pcm and audio codec like mp3 decoder?

**Comment by lgirdwood** [comment]
@muddogxp not today, I would start with a simple definition and hard code some values to get going.

**Comment by dbaluta** [comment]
@paulstelian97 I think we should re-start looking at this as soon as possible. I think we can even start small.
For example, we could find some open source implementation of an mp3 decoder and just add it as a separate component compiled inside the SOF firmware.

So, basically we replace the Volume component from a normal flow with our new processing component. This is just to experiment with the buffers and stuff like that. 

After, that we can go back and see how we integrate the binary.

**Comment by dbaluta** [comment]
@muddogxp when you say 'compressed -> decompressed' do you mean in memory decoding of an mp3 into a wav file for example?

This will be ideal for us, but it needs some deep thinking. We first plan to 'render' mp3 after decompression directly to a DAI interface.

**Comment by muddogxp** [comment]
> @muddogxp not today, I would start with a simple definition and hard code some values to get going.

Looking forward to it.

**Comment by muddogxp** [comment]
> @muddogxp when you say 'compressed -> decompressed' do you mean in memory decoding of an mp3 into a wav file for example?
> 
> This will be ideal for us, but it needs some deep thinking. We first plan to 'render' mp3 after decompression directly to a DAI interface.

Yes, I mean the mp3 decoder in firmware.

**Comment by lgirdwood** [comment]
@paulstelian97 ping - I guess the decoder support is still WIP. Tagging for v1.8

**Comment by paulstelian97** [comment]
> @paulstelian97 ping - I guess the decoder support is still WIP. Tagging for v1.8

I guess sure. Statically linked there is good progress (@dbaluta can you provide more details?) but no work done for dynamic loading yet. Compress interface might be done in 1.7, not sure.

**Comment by paulstelian97** [comment]
Note to myself: Submit an issue instead of this, will send PR when actual progress was made on the dynamic side. Preserving branch...
